clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -4429,9 +4429,9 @@ static void handleSYCLDeviceAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
         << AL << 0 /* static function or anonymous namespace */;
     return;
   }
-  if (isa<CXXMethodDecl>(FD)) {
+  if (FD->isVirtualAsWritten() || FD->isPure()) {
```

Comment:
Do you have a case where the 'isPure' check is required?  I don't know of any case where you can mark a function as pure but not as virtual.  Additionally, I'd suggest just using this function: https://clang.llvm.org/doxygen/classclang_1_1CXXMethodDecl.html#a1f488b7521f08ced46dce05c44cab5c5

Reply:
I agree that using `isVirtual` function seems more elegant. I also can't recall cases when pure function is not virtual.
But then, I'm just curious, why `isVirtual` function also checks both `isVirtualAsWritten` and `isPure` internally. (see https://clang.llvm.org/doxygen/DeclCXX_8h_source.html#l01948 )

Reply:
Ah, because Microsoft.  __interface is a markup in MSVC (https://docs.microsoft.com/en-us/cpp/cpp/interface?view=vs-2019) that makes every function in it pure, but not marked virtual as written :) 

For standards compliant C++, only the 'isvirtualaswritten' and size_overridden_methods checks actually are necessary.